### DISCUSSION TITLE
WebQuery POST doesn't work correctly

### DISCUSSION THREAD

**Discussion by 2000ede** [arendst/Tasmota#20591] [Q&A] [answered]
Hello,
I want to issue a POST command in Tasmota. Unfortunately, I'm probably doing something wrong because it doesn't reach the other side correctly.

`WebQuery http://..../ POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg] data={"serial":"11676", "limit_type":1, "limit_value":5}`

With another system (google script) it works perfectly.

What am I doing wrong?

this works with google script:

`function Limit() {

  var vSN = 11676 
  var user = "admin"
  var pw = "adminadmin"
  var vLimit = "5"

var options = {
  'method' : 'post',
  'payload' : 'data={"serial":"' + vSN + '", "limit_type":1, "limit_value": ' + vLimit +'}',
  'headers' : {"Authorization" : "Basic " + Utilities.base64Encode(user + ':' + pw)}
};
var response = UrlFetchApp.fetch('http://URL……, options);

Logger.log(response);
}
`

![grafik](https://github.com/arendst/Tasmota/assets/37264959/74dda7ca-4a23-4d1c-b10b-789bcf67bfa6)

**Comment by sfromis** [comment]
Without details behind a vague phrase like "it doesn't reach the other side correctly", not easy to see what happens, or what "correctly" even means to you.

One way of debugging could be to compare the result of posting both requests to http://httpbin.org/post
Doing that, my results looks quite reasonable, except for:

I do notice one odd thing in your request, it looks like your payload should be in JSON format,
but nevertheless you have it starting with ` data=` which means that it is not valid JSON,
and not the same as your Google Script example.

In case you did not spot what I meant by "not the same", the difference is that you have an extra space at the start of the payload.

**Reply by 2000ede** [reply]
Thanks for the super fast and detailed answer!
I apologize if I haven't included all the information yet. I'm a Tasmota beginner and not a real programmer. But I'm continuing to work my way forward and it's a lot of fun.

On Google Sheets I get this response:
"Info {"type":"success","message":"Settings saved!","code":1001}"

At Tasmota I get results from my attempts via the console: RSL: RESULT = {"WebQuery":"Done"}

What can and must I do in Tasmota so that I receive more feedback as desired?

**Reply by sfromis** [reply]
First: Did you try fixing the anomaly of the extra space? That could make a difference to your server.

What I meant was the response from directing the POST to `http://httpbin.org/post` instead, as this will provide details of the received request.

If you use a build with `#define USE_WEBSEND_RESPONSE`, the Tasmota response from doing the WebQuery to `httpbin.org/post` should become something like `{"WebQuery":{  "args": {}, ....`

And no, doing a custom build is not hard to do.... https://tasmota.github.io/docs/Compile-your-build/

Anyway, for me doing this command:
`WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg] data={"serial":"11676","limit_type":1,"limit_value":5}`
results in this (formatted for readability)
```
{"WebQuery":{
   "args": {},
   "data": " data={\"serial\":\"11676\",\"limit_type\":1,\"limit_value\":5}",
   "files": {},
   "form": {},
   "headers": {
     "Accept-Encoding": "identity;q=1,chunked;q=0.1,*;q=0",
     "Authorization": "Basic YWRtaW46YWRtaW5hZG1pbg",
     "Content-Length": "55",
     "Host": "httpbin.org",
     "User-Agent": "ESP32HTTPClient",
     "X-Amzn-Trace-Id": "xxxx"
   },
   "json": null,
   "origin": "xx.xx.xx.xx",
   "url": "http://httpbin.org/post"}}
```
The extra space is very visible near the start.

**Comment by 2000ede** [comment]
Thanks!
OK, first an interim status. I'm trying to read more and understand it.

But here is the export:

What strikes me as a layman is that it's under "form" on Google Script and under "data" on Tasmota.
But I can't evaluate it.

**Tasmota:**

18:22:12.441 RSL: RESULT = {"WebQuery":{  "args": {},   "data": " data={\"serial\":\"113182852676\", \"limit_type\":1, \"limit_value\":90}",   "files": {},   "form": {},   "headers": {    "Accept-Encoding": "identity;q=1,chunked;q=0.1,*;q=0",     "Authorization": "Basic YWRtaW46YWRtaW5hZG1pbg",     "Content-Length": "65",     "Host": "httpbin.org",     "User-Agent": "ESP32HTTPClient",     "X-Amzn-Trace-Id": "Root=1-65b298c4-7a76b921312fd1df48b2aa65"  },   "json": null,   "origin": "91.249.56.xxx",   "url": "http://httpbin.org/post"}}

**google:**
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "data": "{\"serial\":\"113182852676\", \"limit_type\":1, \"limit_value\": 30}"
  }, 
  "headers": {
    "Accept-Encoding": "gzip, deflate, br", 
    "Authorization": "Basic YWRtaW46YWRtaW5hZG1pbg==", 
    "Content-Length": "65", 
    "Content-Type": "application/x-www-form-urlencoded", 
    "Host": "httpbin.org", 
    "User-Agent": "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd-928555oE)", 
    "X-Amzn-Trace-Id": "Root=1-65b296bf-4e6b4fcfd5"
  }, 
  "json": null, 
  "origin": "37.138.69.xxx, 107.178.193.xxx", 
  "url": "http://httpbin.org/post"
}

**Reply by sfromis** [reply]
What you post shows that when using Google Scripts, the POST was not done as a simple string, but instead with `Content-Type: application/x-www-form-urlencoded`. To make the request from Tasmota have the same decoration, you should specificy this among the headers in the request.

BTW. Since I see that you're using ESP32, you could get a lot better coding features, including http client requests, using the Berry language. Of course the value of that depends on what you're doing, but it is a big step up compared to doing stuff in rules.

**Reply by barbudor** [reply]
@2000ede Don't add  `data=` in your command
With a post, the payload provide is automatically going as data
```
WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg] {"serial":"11676","limit_type":1,"limit_value":5}
```
other wise `data=`is part of what is snet and I bet the other end dosn't understand that

**Reply by barbudor** [reply]
And by the way, I hope that username:password in the Authorization header are not the real one

Tell me  you haven't used so simple password ??????
If so, now you are good to change them

**Reply by sfromis** [reply]
Yeah, I also observed the oddity of that `data=` at the start of the payload, but since it seems to work when doing the test from Google Script, it appears to be acceptable to the server.

@barbudor - Do note that the format you suggests has a blank at the start of the payload. What I'd suggest would instead be
```
WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg|Content-Type: application/x-www-form-urlencoded]data={"serial":"11676","limit_type":1,"limit_value":5}
``` 
with the already mentioned Content-Type header, and no blank after the `]` terminating the headers. That blank is what I already hinted at a few times :wink:

**Reply by 2000ede** [reply]
Thanks for your help!

I've tried everything for more than a hour now :-) it still doesn't work. I'll try again tomorrow.
Currently it looks like this (please forgive me if I did it wrong):

`WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg|Content-Type: application/x-www-form-urlencoded]{"serial":"11676","limit_type":1,"limit_value":5}`

![grafik](https://github.com/arendst/Tasmota/assets/37264959/47c3fec5-f479-436b-b268-9284b260202b)

**Reply by sfromis** [reply]
Well, it seems that your server depends on that `data=` being part of the payload.
When including that, it seems to work as you expect.
```
WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg|Content-Type: application/x-www-form-urlencoded]data={"serial":"11676","limit_type":1,"limit_value":5}
``` 
results in:
```
{"WebQuery":{
  "args": {},
  "data": "",
  "files": {},
  "form": {
    "data": "{\"serial\":\"11676\",\"limit_type\":1,\"limit_value\":5}"
  },
  "headers": {
    "Accept-Encoding": "identity;q=1,chunked;q=0.1,*;q=0",
    "Authorization": "Basic YWRtaW46YWRtaW5hZG1pbg",
    "Content-Length": "54",
    "Content-Type": "application/x-www-form-urlencoded",
    "Host": "httpbin.org",
    "User-Agent": "ESP32HTTPClient",
    "X-Amzn-Trace-Id": "xxxx"  },
  "json": null,
  "origin": "xx.xx.xx.xx",
  "url": "http://httpbin.org/post"}}
```
And this is the same layout as you post about getting with Google Script. It seems that `data=` is part of the mix expected by your server, even though it is an odd format. Not sure how much httpbin is parsing there, I'd have preferred it to be posting raw data....

**Reply by 2000ede** [reply]
Hello sfromis,
first of all, thank you for your support and patience!

unfortunately it still doesn't work. The command changes a limit on another ESP32 device (not tasmota). The limit is changed via Google script, but unfortunately still not via Tasmota.

WebQuery http://httpbin.org/post POST [Authorization: Basic YWRtaW46YWRtaW5hZG1pbg|Content-Type: application/x-www-form-urlencoded]data={"serial":"11676","limit_type":1,"limit_value":5}

maybe the equal sign by Authorization? (Unfortunately I don't have enough knowledge)

![grafik](https://github.com/arendst/Tasmota/assets/37264959/89c0420a-1393-4d19-a9cc-ac5a97c7e889)


google script:
![grafik](https://github.com/arendst/Tasmota/assets/37264959/e4f4aec4-ab5d-4990-b5ba-e08b927fbe8f)

**Reply by sfromis** [reply]
Well, the difference with `==` at the end of the `Authorization` header from Google Script is that if you want to have those from Tasmota, you'll have to type them in the text, as you are already doing the base64 encoding yourself.

And I already recommended using the Berry language available on ESP32, where you do not have to make the base64 encoding beforehand (with the risk of "forgetting" some of it), like this:
```
var user='admin'
var password='adminadmin'
var authentication=bytes().fromstring(user+':'+password).tob64()
```

**Reply by 2000ede** [reply]
I adjusted it, but it still doesn't work. I only see this difference. 
(or I make another mistake as a beginner.)

(I'll take a look at Berry later, but all my code is in script. I don't have any knowledge of Berry yet).

![grafik](https://github.com/arendst/Tasmota/assets/37264959/1febcb0c-5b05-47af-9633-6ab41dd24790)

**Reply by sfromis** [reply]
If your server has very particular expectations there, you can provide that header along with the existing two headers, to have more added to the header which will be included in the http request.

As `Accept-Encoding` is about what type of encoding the client supports for the response, it would be "atypical" for the server to have very particular expectations there, but if the server chooses to have the response compressed with Brotli (br), Tasmota would not support that.

More generally, if your server is moaning about something, it "should" have an error log with details.

**Reply by 2000ede** [reply]
Thanks for helping.

Unfortunately I didn't manage to do it. As a beginner, it took me hours, even days. I'm giving up for now and will try the detour via Berry. I'm not familiar with this yet, but the command works there.
Thank you again for your commitment, especially sfromis

**Comment by 2000ede** [comment, marked as answer]
Workaround with Berry
